core/trino-main/src/main/java/io/trino/execution/scheduler/FaultTolerantStageScheduler.java

```diff
@@ -236,8 +242,8 @@ public synchronized void schedule()
                     sourceBucketNodeMap);
         }
 
-        while (!queuedPartitions.isEmpty() || !taskSource.isFinished()) {
-            while (queuedPartitions.isEmpty() && !taskSource.isFinished()) {
+        while (!pendingPartitions.isEmpty() || !queuedPartitions.isEmpty() || !taskSource.isFinished()) {
+            while (queuedPartitions.isEmpty() && pendingPartitions.size() != maxTasksWaitingForNodePerStage && !taskSource.isFinished()) {
```

Comment:
Be more defensive?
```suggestion
            while (queuedPartitions.isEmpty() && pendingPartitions.size() < maxTasksWaitingForNodePerStage && !taskSource.isFinished()) {
```

Reply:
make sense. Also changed below `pendingPartitions.size() == maxTasksWaitingForNodePerStage` to `pendingPartitions.size() >= maxTasksWaitingForNodePerStage`